Bulk Ingest with sample sets (#8325)
broadinstitute/gatk: scripts/variantstore/beta_docs/gvs-bulk-ingest-instructions.md

# Bulk Ingest Instructions

The bulk ingest workflow has one distinct difference from it's non-bulk counterpart and required several more input parameters
It can load in more than 10k samples at a time
We do this by directly mapping the sample ids and their gcp storage path locations and loading them that way.

> In order to do this, we need some additional information from you the user.

Reviewer: Most of the new inputs seem to be optional so they wouldn't strictly be needed

Since we ask for so many additional inputs, we have tried to streamline the process in two ways.
The first is that we have defaults of the most common values that are input automatically. They are listed below. 
The second is that we use the information we have to the best of our abilities to make educated guesses about what columns you may want to use

The logic for these guesses is also below.

Since you are planning to use a bulk ingest workflow, we recommend that you use a sample set to load your data

> ## TODO: fellow variant team members, should this be required?!?!? Why would they bulk ingest and NOT use a sample set?

Author: Kylee says we want ALL users to be able to use bulk blech

---

#### Input descriptions

The table below describes the GVS bulk ingest variables:

| Input variable name         | Description                                                                                                                                                                | Type |
|-----------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------| --- |
| data_table_name             | (Optional) The name of the data table; This table hold the GVCFS to be ingested; `sample` is most likely the name of the data table in Terra.                              | String |
| entity_id_column_name       | (Optional) User defined column id name; `sample_id` is the recommended name--`research_id` is also commonly used.                                                          | String |
| vcf_files_column_name       | (Optional) Column that contains the cloud paths to the sample GVCF files                                                                                                   | String |

> | vcf_index_files_column_name | (Optional) Column that contains the cloud paths to the sample GVCF index files                                                                                             | String |

Reviewer: ```suggestion
| vcf_index_files_column_name | (Optional) Column that contains the GCS paths to the sample GVCF index files                                                                                             | String |
```

| bq_project_id               | Name of the Google project that contains the BigQuery dataset.                                                                                                             | String |
| call_set_identifier         | Used to name the filter model, BigQuery extract tables, and final joint VCF shards; should begin with a letter; valid characters include A-z, 0-9, “.”, “,”, “-“, and “_”. | String |

## TODO: do we care about documenting the additional optional input values?

        # File? gatk_override

        # File interval_list = "gs://gcp-public-data--broad-references/hg38/v0/wgs_calling_regions.hg38.noCentromeres.noTelomeres.interval_list"

        # String drop_state = "NONE"

        # The larger the `load_data_batch_size` the greater the probability of preemptions and non-retryable BigQuery errors,
        # so if specifying `load_data_batch_size`, adjust preemptible and maxretries accordingly. Or just take the defaults, as those should work fine in most cases.
        Int? load_data_batch_size
        Int? load_data_preemptible_override

>         Int? load_data_maxretries_override

Author: oops! Nice catch! I think I was trying to remind myself of the other params to potentially document

> # Bulk Ingest Instructions

Reviewer 2: I'd prefer that these instructions are integrated into the main beta docs instead of a new separate document.

Author: that feels like a new ticket though--specifically adding this to beta documentation

Reviewer 3: I was planning to create a new ticket for taking these instructions and using them to update our various "docs with directions."

Reviewer 2: Got it - in that case can we just move this doc out of the beta_docs folder in the meantime? Thanks!

| entity_id_column_name       | (Optional) User defined column id name; `sample_id` is the recommended name;`research_id` is also commonly used.                                                           | String |                                                         | String |
| vcf_files_column_name       | (Optional) Column that contains the GCS paths to the sample GVCF files.                                                                                                    | String |
| vcf_index_files_column_name | (Optional) Column that contains the GCS paths to the sample GVCF index files.                                                                                              | String |
| sample_set_name             | (Optional) The name of the set of samples / entities.                                                                                                                      | String |
| dataset_name                | Name of the BigQuery dataset used to hold input samples, filtering model data, and other tables created during the workflow.                                               | String |

> | bq_project_id               | Name of the Google project that contains the BigQuery dataset.                                                                                                             | String |

Reviewer 4: ```suggestion
| project_id               | Name of the Google project that contains the BigQuery dataset.                                                                                                             | String |
```
It's 'project_id' everywhere else in our workflows.

If you know the values that you want for these 4, feel free to fill them in, but it is not necessary.
If the user does not supply all 4 values, the system uses a mix of default values and heuristics to determine these parameters.
The first parameter, `data_table_name`, is the name of the data entity and is overwhelmingly labelled "sample". This is the default that is used in the system. User documentation likewise suggests this term as does the Beta workspace.
The second parameter, `entity_id_column_name`, dictates the names of the samples as they will show up in GVS. Some advanced users prefer to use a custom column. This defaults to the value of `data_table_name` + "_id" which is most likely "sample_id".

The next two parameters, `vcf_files_column_name` and `vcf_index_files_column_name`, name the two columns that track the GCP locations of the GVCFs. 
If they are not input by the user, the system will look at all the possible columns and a subset of the data in each to predict which columns have the sample paths. The system looks for paths ending in `vcf` or similar.

> The system will also validate that any user input for these parameters fits with the expected relationships among them. 

Reviewer 3: also, I think this sentence should be a part of the paragraph above it

# Bulk Ingest Instructions

The bulk ingest workflow has one distinct difference from its non-bulk counterpart and requires several more input parameters.
It can load in more than 10k samples at a time by directly mapping sample ids to their GCS storage path locations.
In order to do this, the bulk ingest workflow needs to collect some additional information from the user.

Since we ask for so many additional inputs, we have tried to streamline the process in two ways.
The first is that we have defaults of the most common values that are input automatically. These are listed below. 

> The second is that we use the information we have to the best of our abilities to make educated guesses about what columns you may want to use

Reviewer: Sentences could use periods at the end throughout.

| sample_set_name             | (Optional) The name of the set of samples / entities                                                                                                                       | String |
| dataset_name                | Name of the BigQuery dataset used to hold input samples, filtering model data, and other tables created during the workflow.                                               | String |
| bq_project_id               | Name of the Google project that contains the BigQuery dataset.                                                                                                             | String |
| call_set_identifier         | Used to name the filter model, BigQuery extract tables, and final joint VCF shards. Should begin with a letter, valid characters include A-z, 0-9, “.”, “,”, “-“, and “_”. | String |

The first four parameters are labelled as optional, but the pipeline needs a value for each of them to process the data. This is because the system can likely determine the four values it needs.

> if you know the values that you want for these 4, feel free to fill them in, but it is not necessary.

Reviewer: ```suggestion
If you know the values that you want for these 4, feel free to fill them in, but it is not necessary.
```

If the user does not supply all 4 values, the system uses a mix of default values and heuristics to determine these parameters.

> The first parameter, `data_table_name`, is the name of the data entity and is overwhelmingly labelled "sample". This is the default that is used in the system. User documentation likewise suggests this term as does the Beta workspace.

Reviewer: ```suggestion
The first parameter, `data_table_name`, is the name of the data entity and is usually labelled "sample". This is the default that is used in the system. User documentation likewise suggests this term as does the Beta workspace.
```

> The second parameter, `entity_id_column_name`, dictates the sames of the samples as they will show up in GVS and GVS outputs. Some advanced users prefer to use a custom column. This defaults to the value of `data_table_name` + "_id" which is most likely "sample_id"

Reviewer: ```suggestion
The second parameter, `entity_id_column_name`, dictates the names of the samples as they will show up in GVS. Some advanced users prefer to use a custom column. This defaults to the value of `data_table_name` + "_id" which is most likely "sample_id"
```

> The next two parameters, `vcf_files_column_name` and `vcf_index_files_column_name`, are used to understand the GCP locations of the GVCFs. 

Author: I dont know!!!

If they are not input by the user, the system will look at all the possible columns and a subset of the data in each to predict which columns have the sample paths. The system looks for paths ending in `vcf` or similar.

> The system will also validate that any user input for these parameters fit with the expected relationships between them. 

Reviewer: ```suggestion
The system will also validate that any user input for these parameters fits with the expected relationships among them. 
```

The workflow is configured to use hg38 (aka GRCh38) as the reference genome.

#### Sample data

The [GVS beta workspace](https://app.terra.bio/#workspaces/gvs-prod/Genomic_Variant_Store_Beta) has been configured with example reblocked GVCF files that you can use to test the workflow.

The GVS workflow takes in reblocked single sample GVCF files and their corresponding index files as `input_vcfs` and `input_vcf_indexes`, respectively. 

In the Beta user workflow, the value examples are as follows:
the samples_table_name value is `sample`
the entity_id_column_name value is `sample_id`
the vcf_files_column_name value is `gvcf`
the vcf_index_files_column_name value is `gvcf_index`
the sample_set_name value is `sample_set_name`

### Import sample GVCF files

This step validates that sample GVCF files contain required annotations and loads samples into BigQuery tables. 
Once the samples are loaded into BQ, the next step can be run, which creates a table of Alternate Alleles. This table can be used to create a filter or to enable search

The next recommended wdl to run is:

> [GvsPopulateAltAllele](https://github.com/broadinstitute/gatk/blob/ah_var_store/scripts/variantstore/wdl/GvsPopulateAltAllele.wdl) 

Reviewer: not sure I understand why this bit is here
